src/Microsoft.AspNetCore.Mvc.DataAnnotations/Internal/DataAnnotationsMetadataProvider.cs

```diff
@@ -176,15 +176,22 @@ public void CreateDisplayMetadata(DisplayMetadataProviderContext context)
 
                 // EnumDisplayNamesAndValues and EnumNamesAndValues
                 //
-                // Order EnumDisplayNamesAndValues to match Enum.GetNames(). That method orders by absolute value,
-                // then its behavior is undefined (but hopefully stable). Add to EnumNamesAndValues in same order but
-                // Dictionary does not guarantee order will be preserved.
+                // Order EnumDisplayNamesAndValues by DisplayAttribute.Order, then by the order of Enum.GetNames(). 
+                // That method orders by absolute value, then its behavior is undefined (but hopefully stable). 
+                // Add to EnumNamesAndValues in same order but Dictionary does not guarantee order will be preserved.
+
                 var groupedDisplayNamesAndValues = new List<KeyValuePair<EnumGroupAndName, string>>();
                 var namesAndValues = new Dictionary<string, string>();
                 var enumLocalizer = _stringLocalizerFactory?.Create(underlyingType);
-                foreach (var name in Enum.GetNames(underlyingType))
+                var enumNames = Enum.GetNames(underlyingType);
+
+                var enumFields = from i in Enumerable.Range(0, enumNames.Length)
```

Comment:
I think we generally avoid C# LINQ syntax and instead prefer calling the .NET APIs. It makes it a lot clearer to see the cost of the code.